File: src/utils/report_builder.py
#!/usr/bin/env python3
"""
Build text reports for completed works.
"""

from collections import defaultdict
from datetime import datetime
from pathlib import Path

from utils.db_storage import default_reports_dir


def _format_money(value):
    return f"{float(value):.2f}"


def _work_line(work):
    price = float(work.get("price", 0)) / 100.0
    quantity = float(work.get("quantity", 1))
    total = price * quantity
    unit = work.get("unit", "")
    unit_suffix = f" {unit}" if unit else ""
    return (
        f"- {work.get('name', '')}: {quantity:g}{unit_suffix} x "
        f"{_format_money(price)} = {_format_money(total)}"
    ), total


def _append_work_lines(lines, works):
    total = 0.0
    for work in works:
        line, line_total = _work_line(work)
        lines.append(line)
        total += line_total
    return total


def build_work_report(personal_info, client, object_data, works, options=None):
    """
    Build plain-text report body.

    Args:
        personal_info: Executor profile text from settings
        client: Client dict
        object_data: Object dict with name and address
        works: List of work dicts
        options: Report layout options dict
    """
    options = options or {}
    lines = []

    if options.get("include_report_header", True):
        header_text = (options.get("report_header_text") or "").strip()
        if not header_text:
            header_text = "ОТЧЁТ О ПРОДЕЛАННЫХ РАБОТАХ"

        header_lines = header_text.splitlines() or [header_text]
        lines.extend(header_lines)
        separator_len = max(len(line) for line in header_lines)
        lines.append("=" * min(separator_len, 40))
        lines.append("")

    if options.get("include_report_date", True):
        now = datetime.now().strftime("%d.%m.%Y %H:%M")
        lines.append(f"Дата формирования: {now}")
        lines.append("")

    if options.get("include_personal_info", True):
        lines.append("Исполнитель:")
        lines.append((personal_info or "").strip() or "—")
        lines.append("")

    if options.get("include_client_name", True):
        lines.append(f"Заказчик: {client.get('name', '')}")

    if options.get("include_client_address", True):
        address = (object_data or {}).get("address") or client.get("address", "")
        if address:
            lines.append(f"Адрес: {address}")
        object_name = (object_data or {}).get("name", "")
        if object_name:
            lines.append(f"Объект: {object_name}")

    if options.get("include_client_name", True) or options.get("include_client_address", True):
        lines.append("")

    lines.extend(["Перечень работ:", "-" * 40])

    if not works:
        lines.append("Нет записей о выполненных работах.")
    else:
        total = 0.0
        if options.get("group_by_subobjects", True):
            groups = defaultdict(list)
            for work in works:
                key = (work.get("subobject_name") or "").strip() or "Без субобъекта"
                groups[key].append(work)

            for subobject_name in sorted(groups.keys(), key=str.casefold):
                group_works = groups[subobject_name]
                lines.append("")
                lines.append(subobject_name + ":")
                total += _append_work_lines(lines, group_works)
        else:
            lines.append("")
            for index, work in enumerate(works, start=1):
                line, line_total = _work_line(work)
                lines.append(f"{index}. {line[2:]}")
                total += line_total

        lines.extend(["", f"Итого: {_format_money(total)}"])

    lines.append("")
    lines.append("=" * 40)
    return "\n".join(lines)


def save_work_report(personal_info, client, object_data, works, options=None, file_path=""):
    """Save report text to file and return path and content."""
    content = build_work_report(personal_info, client, object_data, works, options)
    reports_dir = default_reports_dir()
    reports_dir.mkdir(parents=True, exist_ok=True)

    if file_path:
        target = Path(file_path).expanduser().resolve()
    else:
        safe_name = "".join(
            ch if ch.isalnum() else "_" for ch in client.get("name", "client")
        )
        stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        target = reports_dir / f"report_{safe_name}_{stamp}.txt"

    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(content, encoding="utf-8")
    return str(target), content

File: cpp/utils/ReportBuilder.cpp
#include "ReportBuilder.h"

#include "DatabaseStorage.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QMap>
#include <algorithm>

namespace workorder
{

namespace
{

bool optionEnabled(const StringMap &options, const char *key, bool defaultValue = true)
{
    if(!options.contains(QLatin1String(key)))
        return defaultValue;
    return options.value(QLatin1String(key)).toBool();
}

QString safeClientName(const StringMap &client)
{
    const QString name = client.value("name").toString();
    QString safe;
    safe.reserve(name.size());
    for(const QChar ch : name)
    {
        if(ch.isLetterOrNumber())
            safe.append(ch);
        else
            safe.append('_');
    }
    if(safe.isEmpty())
        safe = QStringLiteral("client");
    return safe;
}

} // namespace

QString ReportBuilder::formatMoney(double value)
{
    return QString::number(value, 'f', 2);
}

QPair<QString, double> ReportBuilder::workLine(const StringMap &work, bool includeCoefficients)
{
    const double price = work.value("price").toDouble() / 100.0;
    const double quantity = work.value("quantity", 1).toDouble();
    const int percentSum = work.value("percent_sum", 100).toInt();
    const double multiplier = percentSum / 100.0;
    const double displayPrice = includeCoefficients ? price : price * multiplier;
    const double total = price * quantity * multiplier;
    const QString unit = work.value("unit").toString();
    const QString unitSuffix = unit.isEmpty() ? QString() : QStringLiteral(" %1").arg(unit);

    QString name = work.value("name").toString();
    if(includeCoefficients)
    {
        const QString coefficients = work.value("coefficients").toString().trimmed();
        if(!coefficients.isEmpty())
            name += QStringLiteral(" (%1)").arg(coefficients);
    }

    QString line;
    if(includeCoefficients)
    {
        line = QStringLiteral("- %1: %2%3 x %4 x %5 = %6")
            .arg(name)
            .arg(quantity, 0, 'g', 12)
            .arg(unitSuffix)
            .arg(formatMoney(price))
            .arg(QString::number(multiplier, 'g', 12))
            .arg(formatMoney(total));
    }
    else
    {
        line = QStringLiteral("- %1: %2%3 x %4 = %5")
            .arg(name)
            .arg(quantity, 0, 'g', 12)
            .arg(unitSuffix)
            .arg(formatMoney(displayPrice))
            .arg(formatMoney(total));
    }
    return {line, total};
}

double ReportBuilder::appendWorkLines(QStringList &lines, const StringMapList &works, bool includeCoefficients)
{
    double total = 0.0;
    for(const QVariant &item : works)
    {
        const StringMap work = item.toMap();
        const QPair<QString, double> line = workLine(work, includeCoefficients);
        lines.append(line.first);
        total += line.second;
    }
    return total;
}

QString ReportBuilder::buildWorkReport(const QString &personalInfo, const StringMap &client,
                                       const StringMap &objectData, const StringMapList &works,
                                       const StringMap &options)
{
    QStringList lines;

    if(optionEnabled(options, "include_report_header"))
    {
        QString headerText = options.value("report_header_text").toString().trimmed();
        if(headerText.isEmpty())
            headerText = QStringLiteral("ОТЧЁТ О ПРОДЕЛАННЫХ РАБОТАХ");

        const QStringList headerLines = headerText.split('\n');
        if(headerLines.isEmpty())
            lines.append(headerText);
        else
            lines.append(headerLines);

        int separatorLen = 0;
        for(const QString &line : lines)
            separatorLen = qMax(separatorLen, line.size());
        lines.append(QString(separatorLen > 40 ? 40 : separatorLen, '='));
        lines.append(QString());
    }

    if(optionEnabled(options, "include_report_date"))
    {
        lines.append(QStringLiteral("Дата формирования: %1")
            .arg(QDateTime::currentDateTime().toString("dd.MM.yyyy HH:mm")));
        lines.append(QString());
    }

    if(optionEnabled(options, "include_personal_info"))
    {
        lines.append(QStringLiteral("Исполнитель:"));
        const QString info = personalInfo.trimmed();
        if(info.isEmpty())
        {
            lines.append(QStringLiteral("—"));
        }
        else
        {
            const QString normalized = QString(info).replace(QLatin1String("\r\n"), QLatin1String("\n"))
                                                    .replace(QLatin1Char('\r'), QLatin1Char('\n'));
            const QStringList infoLines = normalized.split(QLatin1Char('\n'));
            for(const QString &line : infoLines)
                lines.append(line);
        }
        lines.append(QString());
    }

    if(optionEnabled(options, "include_client_name"))
        lines.append(QStringLiteral("Заказчик: %1").arg(client.value("name").toString()));

    if(optionEnabled(options, "include_client_address"))
    {
        QString address = objectData.value("address").toString();
        if(address.isEmpty())
            address = client.value("address").toString();
        if(!address.isEmpty())
            lines.append(QStringLiteral("Адрес: %1").arg(address));

        const QString objectName = objectData.value("name").toString();
        if(!objectName.isEmpty())
            lines.append(QStringLiteral("Объект: %1").arg(objectName));
    }

    if(optionEnabled(options, "include_client_name") || optionEnabled(options, "include_client_address"))
        lines.append(QString());

    lines.append(QStringLiteral("Перечень работ:"));
    lines.append(QString(40, '-'));

    if(works.isEmpty())
    {
        lines.append(QStringLiteral("Нет записей о выполненных работах."));
    }
    else
    {
        double total = 0.0;
        const bool includeCoefficients = optionEnabled(options, "include_coefficients");
        if(optionEnabled(options, "group_by_subobjects"))
        {
            QMap<QString, StringMapList> groups;
            for(const QVariant &item : works)
            {
                const StringMap work = item.toMap();
                QString key = work.value("subobject_name").toString().trimmed();
                if(key.isEmpty())
                    key = QStringLiteral("Без субобъекта");
                groups[key].append(work);
            }

            QStringList keys = groups.keys();
            std::sort(keys.begin(), keys.end(), [](const QString &a, const QString &b)
            {
                return a.compare(b, Qt::CaseInsensitive) < 0;
            });

            for(const QString &subobjectName : keys)
            {
                lines.append(QString());
                lines.append(subobjectName + ':');
                total += appendWorkLines(lines, groups.value(subobjectName), includeCoefficients);
            }
        }
        else
        {
            lines.append(QString());
            int index = 1;
            for(const QVariant &item : works)
            {
                const StringMap work = item.toMap();
                const QPair<QString, double> line = workLine(work, includeCoefficients);
                lines.append(QStringLiteral("%1. %2").arg(index++).arg(line.first.mid(2)));
                total += line.second;
            }
        }

        lines.append(QString());
        lines.append(QStringLiteral("Итого: %1").arg(formatMoney(total)));
    }

    lines.append(QString());
    lines.append(QString(40, '='));
    return lines.join('\n');
}

QPair<QString, QString> ReportBuilder::saveWorkReport(const QString &personalInfo, const StringMap &client,
                                                        const StringMap &objectData, const StringMapList &works,
                                                        const StringMap &options, const QString &filePath)
{
    const QString content = buildWorkReport(personalInfo, client, objectData, works, options);
    const QString reportsDir = DatabaseStorage::defaultReportsDir();
    QDir().mkpath(reportsDir);

    QString target;
    if(!filePath.isEmpty())
    {
        target = QFileInfo(filePath).absoluteFilePath();
    }
    else
    {
        const QString stamp = QDateTime::currentDateTime().toString("yyyyMMdd_HHmmss");
        target = QDir(reportsDir).filePath(QStringLiteral("report_%1_%2.txt").arg(safeClientName(client), stamp));
    }

    QDir().mkpath(QFileInfo(target).absolutePath());
    QFile file(target);
    if(file.open(QIODevice::WriteOnly | QIODevice::Text))
        file.write(content.toUtf8());

    return {target, content};
}

} // namespace workorder
